src/data/finvec.lean

```diff
@@ -0,0 +1,193 @@
+/-
+Copyright (c) 2020 Reid Barton. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Reid Barton
+-/
+
+import data.equiv.fin
+
+/-!
+# Finite vectors, implemented as functions on `fin n`.
+
+The main advantage of `finvec` over alternatives like `vector` is that
+extracting coordinates (components) of a `finvec` is just function application,
+making it easy to reason about things like permuting coordinates.
+
+## Main definitions
+
+* `finvec n α` is by definition the type `fin n → α`. We think of it as
+  the `n`-fold product `αⁿ` and imagine a value `x : finvec n α` as
+  representing a vector `(x 0, ..., x (n-1))`.
+* `finvec.append`, `finvec.left` and `finvec.right` together exhibit the
+  isomorphism `αⁿ⁺ᵐ ≅ αⁿ × αᵐ`.
+* `finvec.snoc`, `finvec.init` and `finvec.last` likewise exhibit the
+  isomorphism `αⁿ⁺¹ ≅ αⁿ × α`.
+-/
+
+/-- A (homogeneous) "vector" of `n` `α`s, implemented as a function from `fin n`.
+The `j`th component of such a vector `x` (where `j : fin n`) is simply `x j`. -/
+def finvec (n : ℕ) (α : Type*) : Type* := fin n → α
```

Comment:
`functor (finvec n)` also seems useful.  But consistency is also nice.

Reply:
Sure, I don't have a strong opinion about the argument order. The `vector` order looks a bit backwards to me because a `finvec` is really a function, but does have the advantage of matching the reading order of `R^n` in math.